xtm/xtm_api.cc

```diff
@@ -0,0 +1,249 @@
+#else
+	queue->readfd = filedes[0];
+	queue->writefd = filedes[1];
+#endif
+
+	if (fcntl(queue->readfd, F_SETFL, O_NONBLOCK) < 0 ||
+	    fcntl(queue->writefd, F_SETFL, O_NONBLOCK) < 0)
+		goto close_fd;
+	if (queue->queue.xtm_scsp_queue_init(size) < 0) {
+		errno = EINVAL;
+		goto close_fd;
+	}
+	return queue;
+
+close_fd:
+	close(queue->readfd);
+	if (queue->readfd != queue->writefd)
+		close(queue->writefd);
+free_queue:
+	free(queue);
+	return NULL;
+};
+
+int
+xtm_queue_delete(struct xtm_queue *queue)
+{
+	int rc = 0;
+	if (close(queue->readfd) < 0)
+		rc = -1;
+	if (queue->readfd != queue->writefd && close(queue->writefd) < 0)
+		rc = -1;
+	free(queue);
+	return rc;
+}
+
+int
+xtm_msg_probe(struct xtm_queue *queue)
+{
+	if (queue->queue.xtm_scsp_queue_free_count() == 0) {
+		errno = ENOBUFS;
+		return -1;
+	}
+	return 0;
+}
+
+unsigned
+xtm_msg_count(struct xtm_queue *queue)
+{
+	return queue->queue.xtm_scsp_queue_count();
+}
+
+int
+xtm_msg_notify(struct xtm_queue *queue)
+{
+	uint64_t tmp = 1;
+	ssize_t cnt;
+	/*
+	 * We must write 8 byte value, because for linux we
+	 * used eventfd, which require to write >= 8 byte at once.
+	 * Also in case of EINTR we retry to write
+	 */
+	while ((cnt = write(queue->writefd, &tmp, sizeof(tmp))) < 0 &&
+	       errno == EINTR)
+		;
+	return (cnt == sizeof(tmp) ? 0 : -1);
+}
+
+int
+xtm_fun_dispatch(struct xtm_queue *queue, fun_for_dispatch fun,
+		 void *fun_arg, int delayed)
+{
+	union xtm_msg xtm_msg;
+	xtm_msg.fun = fun;
+	xtm_msg.fun_arg = fun_arg;
+	if (queue->queue.xtm_scsp_queue_put(&xtm_msg, 1) != 1) {
+		errno = ENOBUFS;
+		return -1;
+	}
+	if (delayed == 0)
+		return xtm_msg_notify(queue);
+	return 0;
+}
+
+int
+xtm_fd(struct xtm_queue *queue)
+{
+	return queue->readfd;
+}
+
+int
+xtm_fun_invoke(struct xtm_queue *queue, int flushed)
+{
+	if (flushed != 0 && xtm_queue_flush_pipe(queue) < 0)
+		return -1;
+	struct xtm_scsp_queue_read_iterator<xtm_msg> iter;
+	iter.xtm_scsp_queue_iterator_begin(&queue->queue);
+	const union xtm_msg *xtm_msg;
+	unsigned cnt = 0;
+
+	while((xtm_msg = iter.xtm_scsp_queue_iterator_read()) != nullptr) {
+		xtm_msg->fun(xtm_msg->fun_arg);
+		cnt++;
+	}
+	iter.xtm_scsp_queue_iterator_end();
+	return cnt;
+}
+
+int
+xtm_msg_send(struct xtm_queue *queue, void *msg, int delayed)
+{
+	union xtm_msg xtm_msg;
+	xtm_msg.data = msg;
+	if (queue->queue.xtm_scsp_queue_put(&xtm_msg, 1) != 1) {
+		errno = ENOBUFS;
+		return -1;
+	}
+	if (delayed == 0)
+		return xtm_msg_notify(queue);
+	return 0;
+}
+
+int
+xtm_msg_recv(struct xtm_queue *queue, void **msg, unsigned count, int flushed)
+{
+	if (flushed != 0 && xtm_queue_flush_pipe(queue) < 0)
+		return -1;
+	struct xtm_scsp_queue_read_iterator<xtm_msg> iter;
+	iter.xtm_scsp_queue_iterator_begin(&queue->queue);
+	const union xtm_msg *xtm_msg;
+	unsigned cnt = 0;
+
+	while(cnt < count && (xtm_msg = iter.xtm_scsp_queue_iterator_read()) !=
+	      nullptr) {
+		msg[cnt] = xtm_msg->data;
+		cnt++;
```

Comment:
It might be faster to do `++msg`, and in the end return `msg - msg_begin`, where `msg_begin` you would remember in the beginning. Then it would be minus 1 address calculation inside of the loop.

Reply:
Fixed